File: elixir/all-your-base/lib/all_your_base.ex
defmodule AllYourBase do
  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """

  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(_digits, input_base, _output_base) when input_base < 2,
    do: {:error, "input base must be >= 2"}

  def convert(_digits, _input_base, output_base) when output_base < 2,
    do: {:error, "output base must be >= 2"}

  def convert([], _input_base, _output_base), do: {:ok, [0]}

  def convert(digits, input_base, output_base) do
    if not valid_input?(digits, input_base) do
      {:error, "all digits must be >= 0 and < input base"}
    else
      res =
        digits
        |> convert_to_base_10(input_base)
        |> convert_from_base_10(output_base, [])

      {:ok, res}
    end
  end

  defp convert_to_base_10(digits, input_base) do
    digits
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.reduce(0, fn {digit, index}, acc ->
      digit * Integer.pow(input_base, index) + acc
    end)
  end

  defp convert_from_base_10(0, _output_base, []), do: [0]
  defp convert_from_base_10(0, _output_base, res), do: res

  defp convert_from_base_10(number, output_base, res) do
    convert_from_base_10(div(number, output_base), output_base, [rem(number, output_base) | res])
  end

  defp valid_input?(digits, input_base) do
    Enum.all?(digits, &(&1 >= 0 && &1 < input_base))
  end
end

File: erlang/all-your-base/src/all_your_base.erl
-module(all_your_base).
-export([rebase/3]).


rebase(_, InputBase, _) when InputBase < 2 -> {error, "input base must be >= 2"};
rebase(_, _, OutputBase) when OutputBase < 2 -> {error, "output base must be >= 2"};
rebase([], _, _) -> {ok, [0]};
rebase(Digits, InputBase, OutputBase) -> 
    to_base10(lists:reverse(Digits), InputBase, OutputBase, 0, 0).

to_base10([], InputBase, OutputBase, _, Acc) -> from_base10(Acc, OutputBase, []);
to_base10([Digit | _], InputBase, _, _, Acc) when Digit < 0 -> 
    {error, "all digits must satisfy 0 <= d < input base"};
to_base10([Digit | _], InputBase, _, _, Acc) when Digit >= InputBase -> 
    {error, "all digits must satisfy 0 <= d < input base"};
to_base10([Digit | RemDigits], InputBase, OutputBase, Index, Acc) -> 
    to_base10(RemDigits, InputBase, OutputBase, Index + 1, Acc + trunc(Digit * math:pow(InputBase, Index))).

from_base10(0, _, []) -> {ok, [0]};
from_base10(0, _, OutputDigits) -> {ok, OutputDigits};
from_base10(InputValue, OutputBase, OutputDigits) -> 
    from_base10(InputValue div OutputBase, OutputBase, [(InputValue rem OutputBase) | OutputDigits]).
